gordon_janitor_gcp/gce_instance_authority.py

```diff
@@ -0,0 +1,145 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+A GCEInstanceAuthority object retrieves a list of all instances in all
+projects that the ``keyfile`` gives it access to. It is meant to work
+in an asynchronous context and to be executed through the ``start()`` method.
+
+It is also meant to be easy to extend and customize so all methods except
+``start`` can be overridden as long as they follow the interfaces over the
+base methods. For example, if a different project filtering scheme is to be
+used, it's possible to inherit from GCEInstanceAuthority and override
+the ``get_all_projects`` method and ensure it returns a set of project IDs.
+
+To use:
+
+.. code-block:: pycon
+
+    >>> import asyncio
+    >>> rrset_channel = asyncio.Queue()
+    >>> authority = GCEInstanceAuthority(config, rrset_channel)
+    >>> await authority.start()
+    >>> msg = await.channel.get()
+    >>> print(msg)
+    {'zone': 'us-central1-b', 'resourceRecords': [...]}
+"""
+
+import aiohttp
+
+from gordon_janitor_gcp import auth
+from gordon_janitor_gcp import gce_clients
+from gordon_janitor_gcp import msg_generators
+
+
+class GCEInstanceAuthority:
+    """Gordon-Janitor plugin that gathers "state of truth" GCE instance data.
+
+    Args:
+        config (dict): Plugin-specific configuration.
+        rrset_channel (asyncio.Queue): channel to receive instance data
+            messages.
+    """
+    def __init__(self, config, rrset_channel=None, **kwargs):
+        self.config = config
+        self.keyfile = self.config['keyfile']
+        self.rrset_channel = rrset_channel
+
+    async def get_all_projects(self):
+        """Get all active GCE projects.
+
+        Returns:
+            set: active GCE project IDs.
+        """
+        auth_client = auth.GoogleAuthClient(
+            self.keyfile,
+            scopes=['cloud-platform.read-only'],
+            session=self.session)
+        crm_client = gce_clients.CRMClient(auth_client, self.session)
+
+        active_projects = await crm_client.list_all_active_projects()
+        return set(p.get('projectId') for p in active_projects)
+
+    async def get_processed_projects(self):
+        """Get list of GCE project IDs with blacklisted ones filtered out.
+
+        Returns:
+            list: project IDs.
+        """
+        projects = await self.get_all_projects()
+        project_blacklist = set(self.config['project_blacklist'])
+        return sorted(projects - project_blacklist)
+
+    async def get_instance_filter(self):
+        """Get filter string used by instances.aggregatedList endpoint.
+
+        For filter string details, see api endpoint documentation:
+        https://cloud.google.com/compute/docs/reference/latest/instances/aggregatedList
+
+        Returns:
+            str: filter suitable for the 'filter' endpoint parameter.
+        """
+        return ''
```

Comment:
is there a reason to have this as a method, as opposed to as an instance variable?  why does it need to be async?

Reply:
Yeah I was going to ask about this - why is it even needed?

Reply:
And if it is needed, maybe it can be a private method (if it has to be a method)

Reply:
Good q's. I set this up this way so that gordon can read some pods from the config (because pods can be read and understood by humans), contact podservice to translate the pods into google zones (eg. 'pod1' -> 'us-central1-z'), and use that zone to build a query string that's understood by the API endpoint (eg. `zone eq .*/us-central1-z'). It returns an empty string because the pod-zone-filter logic is specific to Spotify and, afaik, shouldn't be in this repo.

If I'm thinking this through right, removing this would mean either having the same logic in a different place or having us configure each instance of the janitor with concrete zone names. Does this sound right?

> And if it is needed, maybe it can be a private method (if it has to be a method)

Good idea, I'll change it to be private.

Reply:
I guess I'm confused at how non-Spotify users are suppose to make use of this filter...